### PR TITLE
Let PoolThreadCache work even if allocation and deallocation Thread a…

### PR DESCRIPTION
**Pull request by normanmaurer**
…re different

Motivation:

PoolThreadCache did only cache allocations if the allocation and deallocation Thread were the same. This is not optimal as often people write from differen thread then the actual EventLoop thread.

Modification:

- Add MpscArrayQueue which was forked from jctools and lightly modified.
- Use MpscArrayQueue for caches and always add buffer back to the cache that belongs to the allocation thread.

Result:

ThreadPoolCache is now also usable and so gives performance improvements when allocation and deallocation thread are different.

Performance when using same thread for allocation and deallocation is noticable worse then before.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin please review.

@nmittler @ejona86 @louiscryan this should also be interesting for grpc :)

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 1]
```diff
@@ -335,27 +338,14 @@ protected void initBuf(
...
 
-        @SuppressWarnings("unchecked")
         MemoryRegionCache(int size) {
-            entries = new Entry[powerOfTwo(size)];
-            for (int i = 0; i < entries.length; i++) {
-                entries[i] = new Entry<T>();
-            }
-            maxUnusedCached = size / 2;
+            this.size = powerOfTwo(size);
+            queue = new MpscArrayQueue<Entry<T>>(this.size);
```
Comment (trustin) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: What happens when `Unsafe` is unavailable?

**Comment by trustin** [comment]
How about making the new classes package-private and adding the factory methods to `PlatformDependent`?

**Comment by normanmaurer** [comment]
@trustin good point let me address both comments

**Comment by normanmaurer** [comment]
@trustin addressed comments... please check

**Comment by trustin** [comment]
See #3835

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 2]
```diff
@@ -335,27 +339,14 @@ protected void initBuf(
...
 
-        @SuppressWarnings("unchecked")
         MemoryRegionCache(int size) {
-            entries = new Entry[powerOfTwo(size)];
-            for (int i = 0; i < entries.length; i++) {
-                entries[i] = new Entry<T>();
-            }
-            maxUnusedCached = size / 2;
+            this.size = powerOfTwo(size);
+            queue = PlatformDependent.newFixedMpscQueue(this.size);
```
Comment (trustin) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: Not really a gain, but maybe this is shorter:

    queue = PlatformDependent.newFixedMpscQueue(this.size = powerOfTwo(size));

**Comment by trustin** [comment]
LGTM. Please consider to take care of #3835 :-)

**Comment by normanmaurer** [comment]
@trustin can we handle #3835 after this is merged ?

**Comment by trustin** [comment]
@normanmaurer Sure!

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java (hunk 2 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: just leave it as it is :)

**Comment by normanmaurer** [comment]
@netkins build

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (f765053ae740e300a6b696840d7dfe5de32afeb3) , 4.1 (81fee66c78b6afad1d92647669c67bcf44f7cbdf) and master (ee1219b6b70878618d382182ea71cc47936ad8de)

**Pull request closed without merging**